File: app/src/main/java/com/qfs/apres/soundfont/Riff.kt
package com.qfs.apres.soundfont

import com.qfs.apres.toUInt
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.io.FileInputStream
import java.io.InputStream

class Riff(private var file_path: String) {
    class InvalidRiff(file_path: String): Exception("$file_path is not a valid Riff")
    class InputStreamClosed : Exception("Input Stream is Closed")
    data class ListChunkHeader(
        val index: Int,
        val tag: String,
        val size: Int,
        val type: String
    )

    data class SubChunkHeader(
        val index: Int,
        val tag: String,
        val size: Int
    )
    lateinit var type_cc: String

    var list_chunks: MutableList<ListChunkHeader> = mutableListOf()
    var sub_chunks: MutableList<List<SubChunkHeader>> = mutableListOf()
    private var _input_stream: InputStream? = null
    private var _input_position: Int = 0
    private var _read_mutex = Mutex()

    init {
        this.with {
            val header_check = this.get_string(0, 4) // fourcc
            if (header_check != "RIFF") {
                this.close_stream()
                throw InvalidRiff(file_path)
            }
            val riff_size = this.get_little_endian(4, 4)
            this.type_cc = this.get_string(8, 4)

            var working_index = 12
            while (working_index < riff_size - 4) {
                val header_index = working_index - 12
                val tag = this.get_string(working_index, 4)
                working_index += 4

                val chunk_size = this.get_little_endian(working_index, 4)
                working_index += 4

                val type = this.get_string(working_index, 4)
                working_index += 4

                this.list_chunks.add(
                    ListChunkHeader(
                        header_index,
                        tag,
                        chunk_size,
                        type
                    )
                )

                val sub_chunk_list: MutableList<SubChunkHeader> = mutableListOf()
                var sub_index = 0
                while (sub_index < chunk_size - 4) {
                    val subchunkheader = SubChunkHeader(
                        index = sub_index + working_index - 12,
                        tag = this.get_string(working_index + sub_index, 4),
                        size = this.get_little_endian(working_index + sub_index + 4, 4)
                    )
                    sub_chunk_list.add(subchunkheader)

                    sub_index += 8 + subchunkheader.size
                }

                working_index += sub_index
                this.sub_chunks.add(sub_chunk_list)
            }
        }
    }

    private fun open_stream() {
        this._input_stream = FileInputStream(this.file_path)
    }

    private fun close_stream() {
        this._input_stream?.close()
        this._input_position = 0
        this._input_stream = null
    }

    fun with(callback: (Riff) -> Unit) {
        runBlocking {
            this@Riff._read_mutex.withLock {
                this@Riff.open_stream()
                try {
                    callback(this@Riff)
                } finally {
                    this@Riff.close_stream()
                }
            }
        }
    }

    fun get_chunk_data(header: SubChunkHeader): ByteArray {
        return this.get_bytes(header.index + 8 + 12, header.size)
    }

    fun get_chunk_data(header: ListChunkHeader): ByteArray {
        return this.get_bytes(header.index + 12 + 12, header.size)
    }

    fun get_list_chunk_data(header: ListChunkHeader, inner_offset: Int? = null, cropped_size: Int? = null): ByteArray {
        // First get the offset of the sub chunk
        var offset = header.index + 8

        // get the *actual* size of the sub chunk
        var size = header.size

        if (inner_offset != null) {
            size -= inner_offset
            offset += inner_offset
        }

        if (cropped_size != null && cropped_size <= size) {
            size = cropped_size
        }

        return this.get_bytes(offset, size)
    }


    fun get_sub_chunk_data(header: SubChunkHeader, inner_offset: Int? = null, cropped_size: Int? = null): ByteArray {
        // First get the offset of the sub chunk
        var offset = header.index + 8

        // get the *actual* size of the sub chunk
        var size = header.size
        if (inner_offset != null) {
            size -= inner_offset
            offset += inner_offset
        }

        if (cropped_size != null && cropped_size <= size) {
            size = cropped_size
        }

        return this.get_bytes(offset, size)
    }

    private fun move_to_offset(offset: Long) {
        val stream: InputStream = this._input_stream ?: throw InputStreamClosed()

        if (this._input_position < offset) {
            stream.skip(offset - this._input_position)

        } else if (this._input_position > offset) {
            this.close_stream()
            this.open_stream()
            this._input_stream?.skip(offset)
        }
        this._input_position = offset.toInt()
    }

    private fun get_bytes(offset: Int, size: Int): ByteArray {
        this.move_to_offset(offset.toLong())
        val output = ByteArray(size)
        this._input_stream?.read(output)
        this._input_position += size
        return output
    }

    private fun get_string(offset: Int, size: Int): String {
        return this.get_bytes(offset, size).toString(Charsets.UTF_8)
    }

    private fun get_little_endian(offset: Int, size: Int): Int {
        var output = 0
        val bytes = this.get_bytes(offset, size)
        for (i in 0 until size) {
            output *= 256
            output += toUInt(bytes[size - 1 - i])
        }
        return output
    }
}

File: app/src/main/cpp/soundfont/Riff.cpp
#include <jni.h>
#include <iostream>
#include <string>
#include <exception>
#include <fstream>
#include <utility>
#include <vector>
#include <string>


class InvalidRiffException: public std::exception {};
struct ListChunkHeader {
    int index;
    std::string tag;
    int size;
    std::string type;
};

struct SubChunkHeader {
    int index;
    std::string tag;
    int size;
};

class Riff {
    std::string path;
    std::vector<ListChunkHeader> list_chunks;
    std::vector<std::vector<SubChunkHeader>> sub_chunks;

    public:
        explicit Riff(std::string _path) {
            this->path = std::move(_path);
            this->list_chunks = {};
            this->sub_chunks = {};

            std::ifstream stream(path);

            std::string header_check = Riff::get_string(&stream, 0, 4); // Fourcc
            if (header_check != "RIFF") {
                stream.close();
                throw InvalidRiffException();
            }

            int riff_size = Riff::get_little_endian(&stream, 4, 4);
            std::string type_cc = Riff::get_string(&stream, 8, 4);

            int working_index = 12;
            while (working_index < riff_size - 4) {
                int header_index = working_index - 12;
                std::string tag = Riff::get_string(&stream, working_index, 4);
                working_index += 4;

                int chunk_size = Riff::get_little_endian(&stream, working_index, 4);
                working_index += 4;

                std::string type = Riff::get_string(&stream, working_index, 4);
                working_index += 4;

                list_chunks.push_back(ListChunkHeader {header_index, tag, chunk_size, type });

                std::vector<SubChunkHeader> working_sub_chunks = {};
                int sub_index = 0;
                while (sub_index < chunk_size - 4) {
                    SubChunkHeader chunkheader = SubChunkHeader {
                        sub_index + working_index - 12,
                        this->get_string(&stream, working_index + sub_index, 4),
                        this->get_little_endian(&stream, working_index + sub_index + 4, 4)
                    };
                    working_sub_chunks.push_back(chunkheader);
                    sub_index += 8 + chunkheader.size;
                }

                working_index += sub_index;
                sub_chunks.push_back(working_sub_chunks);
            }

            stream.close();
        }

        static std::string get_string(std::ifstream* stream, int start, int length) {
            stream->seekg(start); char output[length];
            stream->read(output, length);
            return output;
        }

        static int get_little_endian(std::ifstream* stream, int start, int length) {
            stream->seekg(start);
            int output = 0;
            char v[1];
            for (int i = 0; i < length; i++) {
                output *= 256;
                stream->read(v, 1);
                output += int(v[0]);
            }
            return output;
        }

        static char* get_bytes(std::ifstream* stream, int start, int length) {
            char output[length];
            stream->seekg(start);
            stream->read(output, length);
            return output;
        }

        static char* get_chunk_data(std::ifstream* stream, ListChunkHeader* header) {
            return Riff::get_bytes(stream, header->index + 24, header->size);
        }

        static char* get_sub_chunk_data(std::ifstream* stream, SubChunkHeader* header, std::optional<int> inner_offset, std::optional<int> cropped_size) {
            int offset = header->index + 8;
            int size = header->size;
            if (inner_offset.has_value()) {
                size -= inner_offset.value();
                offset += inner_offset.value();
            }

            if (cropped_size.has_value() and cropped_size.value() <= size) {
                size = cropped_size.value();
            }

            return Riff::get_bytes(stream, offset, size);
        }
};
